### ISSUE TITLE
Vim9: line number in error message is not correct

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6724]
**Describe the bug**

In Vim9 script, when an error is raised from a `:for` loop, the line number is not correct.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        def Func()
            for d in [{}]
                if true
                    eval [][0]
                endif
            endfor
        enddef
        Func()
    EOF
    )

`E684` is raised:

    E684: list index out of range: 0

This is expected.

The error line number is `5`:

    line    5:

This is *not* expected.

**Expected behavior**

The error line number is `3`, because the error comes from the third line in the body of `Func()`.

    eval [][0]

**Environment**

 - Vim version: 8.2 Included patches: 1-1471
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: XTerm(358)

**Additional context**

This makes it harder to debug an issue; we need to find the correct line manually.

---

Here's yet another similar example where the line number is not correct:
```vim
vim9script
def Func()
    for d in [{}]
        if !has_key(d, 'x')
            eval [][0]
        endif
    endfor
enddef
Func()
```
This time, the message prints the error line number `2`:

    line    2:

This is – again – not correct; the error line number is still `3`.

**Comment by lacygoill** [comment]
The issue looks similar to #6708 which was fixed by [8.2.1446](https://github.com/vim/vim/releases/tag/v8.2.1446).

---

It's caused by a regression introduced in [8.2.1463](https://github.com/vim/vim/releases/tag/v8.2.1463).

**Comment by lacygoill** [comment]
In the previous examples, it's quite easy to find out from where the error really comes from.  But in a real code, it's much more difficult.

For example, I've refactored the C omni completion function [here](https://github.com/lacygoill/config/blob/master/.vim/autoload/ccomplete.vim) in Vim9 script.  It worked as expected until a few days ago.  It broke recently because of the regression reported in #6723.  Here is the error message:

    Error detected while processing function ccomplete#Complete:
    line    1:
    E1030: Using a String as a Number

Notice that the message says that the error is on line 1.  But there is no code on line 1; just [a comment](https://github.com/lacygoill/config/blob/a788ceb980f86c4a2619ac3de20b11e39e063c10/.vim/autoload/ccomplete.vim#L7).  The real error comes from [this line](https://github.com/lacygoill/config/blob/a788ceb980f86c4a2619ac3de20b11e39e063c10/.vim/autoload/ccomplete.vim#L197), which has the address 191 in the body of `ccomplete#Complete())` according to `:fu ccomplete#Complete`.  This is far away from 1, and thus hard to find, unless you understand the issue very well.
